embedded-nal-async/src/stack/tcp.rs

```diff
@@ -170,3 +170,45 @@ impl<T: TcpClientStack> TcpClientStack for &mut T {
 		T::close(self, socket)
 	}
 }
+
+/// This trait is implemented by TCP/IP stacks. In contrast to the TcpClientStack trait,
+/// this trait allows creating multiple connections using the same TCP/IP stack.
+///
+/// The trait relies on an associated type, `TcpConnection` which is returned when
+/// creating a TCP connection.
+pub trait TcpClient: embedded_io::Io {
+	/// Type returned when creating a connection.
+	type TcpConnection<'m>: embedded_io::asynch::Read + embedded_io::asynch::Write + Close
+	where
+		Self: 'm;
+
+	/// Options that may be passed to connect.
+	type ConnectOpts<'m>: Default
```

Comment:
This seems like something that would make it extremely hard to use this as a proper abstraction between library crates?

It works for abstracting between a library and user, but in that case one might as well just use provided methods on the library without a trait? 

I can implement this trait for a library crate providing e.g. a driver for a networking chip (where the `ConnectOpts` would be specific for that chip/library), but i cannot use this abstraction in a crate doing networking? E.g. i cannot write an MQTT or HTTP client using this, other than always calling `default()` in which case i don't think it give anything?

Reply:
Yeah, I'm coming to realize that just now as well. I was adding this because I am experimenting with having a TLS client implementing this trait in a 'transparent' way, but I'm running into that specific issue. So I'm thinking maybe it's best to remove the ConnectOpts.